src/iface/ethernet.rs

```diff
@@ -25,13 +24,12 @@ use super::ArpCache;
 /// The network interface logically owns a number of other data structures; to avoid
 /// a dependency on heap allocation, it instead owns a `BorrowMut<[T]>`, which can be
 /// a `&mut [T]`, or `Vec<T>` if a heap is available.
-pub struct Interface<'a, 'b, 'c, 'd: 'a, DeviceT: Device<'d> + 'a> {
+pub struct Interface<'a, 'b, 'c, DeviceT: for<'d> Device<'d> + 'a> {
```

Comment:
Oh, this also works. Sweet, this is the 1st time I find higher-ranker lifetimes useful, personally.

Reply:
Me too :). I think it's even necessary here because `transmit`/`receive` take `&'a mut self` to support tokens with a lifetime. Without the higher-ranked lifetime, `transmit` would always borrow for the fixed `'d` lifetime defined in `Interface`, which means that the borrow lasts for the whole device lifetime. So we could only `transmit` once. With higher-ranked lifetimes, arbitrarily short borrows are possible (because `DeviceT` must implement `Device` for all possible lifetimes).